### PR TITLE
Add upgrade wrapper for `clusterctl` and skeleton for `clusterapi` upgrader

### PR DESCRIPTION
**Pull request by g-gaston**
This is part of #314 

*Description of changes:*
In order to have smaller PRs, splitting this issue into multiple parts
This one adds the `upgrade apply` command wrapper to `clusterctl` and the skeleton for a `clusterapi` upgrader

The idea is that the `clusterapi.Upgrader` is a higher level entity that takes care of comparing versions and determine what has changed and needs to be upgraded. And `clusterctl` just generates the proper command to update those components to those specific versions

The `clusterapi.Upgrader` will probably be called from the upgrade workflow or from `ClusterManager` directly

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by abhay-krishna** [comment]
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/aws/eks-anywhere/pull/376#" title="Author self-approved">g-gaston</a>*

The full list of commands accepted by this bot can be found [here](https://go.k8s.io/bot-commands?repo=aws%2Feks-anywhere).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[OWNERS](https://github.com/aws/eks-anywhere/blob/main/OWNERS)~~ [g-gaston]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

File: pkg/executables/clusterctl.go
Comment (vivek-koppuru) [inline pkg/executables/clusterctl.go:282]: Should we just change for these too?

File: pkg/types/upgrades.go
Comment (vivek-koppuru) [inline pkg/types/upgrades.go:7]: Report makes me think of a list of information in a text based format. What about `ComponentChangeDiff`, `ComponentDisparity`, or `ComponentCompare`

File: pkg/types/upgrades.go
Comment (vivek-koppuru) [inline pkg/types/upgrades.go:9]: How do we plan on using the old version? Can't we just have a hash map instead of this struct and only add to hashmap with component as key and new version as value if there is a change and check based on that?

File: cmd/eksctl-anywhere/cmd/generateclusterconfig.go
Comment (g-gaston) [reply cmd/eksctl-anywhere/cmd/generateclusterconfig.go:62]: If the constants live in the providers then they can't be used from certain packages because they create cyclic dependencies
The other alternative is to duplicate them

What's the concern with mapping?

File: pkg/executables/clusterctl.go
Comment (g-gaston) [reply pkg/executables/clusterctl.go:282]: I didn't because I didn't need those values as constants
But if you feel it's better, I can do it

File: pkg/types/upgrades.go
Comment (g-gaston) [reply pkg/types/upgrades.go:7]: I like `ComponentDiff`, will change

File: pkg/types/upgrades.go
Comment (g-gaston) [reply pkg/types/upgrades.go:9]: Interesting. What's the benefit of using a map? Imo it's way more complex to expand it. And we don't need to do lookups anywhere.

My idea was that every "owner" of components would have a method to return these. That way we can check what needs to change before actually changing it. I think we'll use it for validations/pre-flight checks.
Also, not in scope for this feature, but it could be used to have a `upgrade cluster plan` command

File: pkg/executables/clusterctl.go
Comment (danbudris) [inline pkg/executables/clusterctl.go:312]: nit: use namespace constant for namespace name, e.g. "%s/kubeadm:%s"

File: pkg/executables/clusterctl.go
Comment (danbudris) [inline pkg/executables/clusterctl.go:316]: nit: use namespace constants for namespace name

**Review by danbudris — APPROVED** [review]

**Comment by abhay-krishna** [comment]
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/aws/eks-anywhere/pull/376#pullrequestreview-771626846" title="Approved">danbudris</a>*, *<a href="https://github.com/aws/eks-anywhere/pull/376#" title="Author self-approved">g-gaston</a>*

The full list of commands accepted by this bot can be found [here](https://go.k8s.io/bot-commands?repo=aws%2Feks-anywhere).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[OWNERS](https://github.com/aws/eks-anywhere/blob/main/OWNERS)~~ [danbudris,g-gaston]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

File: pkg/types/upgrades.go
Comment (g-gaston) [reply pkg/types/upgrades.go]: Done

**Comment by abhay-krishna** [comment]
@g-gaston: The following test **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
eks-anywhere-attribution-files-presubmit | c9f464153c75b4bf74f1dd65de97f5c2da8b1e3d | [link](https://prow.eks.amazonaws.com/view/s3/prowpresubmitsdataclusterstack-prowbucket7c73355c-vfwwxd2eb4gp/pr-logs/pull/aws_eks-anywhere/376/eks-anywhere-attribution-files-presubmit/1445473385445330944) | `/test eks-anywhere-attribution-files-presubmit`

[Full PR test history](https://prow.eks.amazonaws.com/pr-history?org=aws&repo=eks-anywhere&pr=376). [Your PR dashboard](https://prow.eks.amazonaws.com/pr?query=is:pr+state:open+author:g-gaston).

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>

**Comment by abhay-krishna** [comment]
@g-gaston: The following tests **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
eks-anywhere-attribution-files-presubmit | c9f464153c75b4bf74f1dd65de97f5c2da8b1e3d | [link](https://prow.eks.amazonaws.com/view/s3/prowpresubmitsdataclusterstack-prowbucket7c73355c-vfwwxd2eb4gp/pr-logs/pull/aws_eks-anywhere/376/eks-anywhere-attribution-files-presubmit/1445473385445330944) | `/test eks-anywhere-attribution-files-presubmit`
eks-anywhere-e2e-presubmit | c9f464153c75b4bf74f1dd65de97f5c2da8b1e3d | [link](https://prow.eks.amazonaws.com/view/s3/prowpresubmitsdataclusterstack-prowbucket7c73355c-vfwwxd2eb4gp/pr-logs/pull/aws_eks-anywhere/376/eks-anywhere-e2e-presubmit/1445473385474691072) | `/test eks-anywhere-e2e-presubmit`

[Full PR test history](https://prow.eks.amazonaws.com/pr-history?org=aws&repo=eks-anywhere&pr=376). [Your PR dashboard](https://prow.eks.amazonaws.com/pr?query=is:pr+state:open+author:g-gaston).

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>

**Comment by abhay-krishna** [comment]
@g-gaston: The following tests **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
eks-anywhere-attribution-files-presubmit | c9f464153c75b4bf74f1dd65de97f5c2da8b1e3d | [link](https://prow.eks.amazonaws.com/view/s3/prowpresubmitsdataclusterstack-prowbucket7c73355c-vfwwxd2eb4gp/pr-logs/pull/aws_eks-anywhere/376/eks-anywhere-attribution-files-presubmit/1445473385445330944) | `/test eks-anywhere-attribution-files-presubmit`
eks-anywhere-e2e-presubmit | c9f464153c75b4bf74f1dd65de97f5c2da8b1e3d | [link](https://prow.eks.amazonaws.com/view/s3/prowpresubmitsdataclusterstack-prowbucket7c73355c-vfwwxd2eb4gp/pr-logs/pull/aws_eks-anywhere/376/eks-anywhere-e2e-presubmit/1445473385474691072) | `/test eks-anywhere-e2e-presubmit`
eks-anywhere-presubmit | c9f464153c75b4bf74f1dd65de97f5c2da8b1e3d | [link](https://prow.eks.amazonaws.com/view/s3/prowpresubmitsdataclusterstack-prowbucket7c73355c-vfwwxd2eb4gp/pr-logs/pull/aws_eks-anywhere/376/eks-anywhere-presubmit/1445473385533411328) | `/test eks-anywhere-presubmit`

[Full PR test history](https://prow.eks.amazonaws.com/pr-history?org=aws&repo=eks-anywhere&pr=376). [Your PR dashboard](https://prow.eks.amazonaws.com/pr?query=is:pr+state:open+author:g-gaston).

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>

**Comment by abhay-krishna** [comment]
@g-gaston: The following tests **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
eks-anywhere-attribution-files-presubmit | c9f464153c75b4bf74f1dd65de97f5c2da8b1e3d | [link](https://prow.eks.amazonaws.com/view/s3/prowpresubmitsdataclusterstack-prowbucket7c73355c-vfwwxd2eb4gp/pr-logs/pull/aws_eks-anywhere/376/eks-anywhere-attribution-files-presubmit/1445473385445330944) | `/test eks-anywhere-attribution-files-presubmit`
eks-anywhere-e2e-presubmit | c9f464153c75b4bf74f1dd65de97f5c2da8b1e3d | [link](https://prow.eks.amazonaws.com/view/s3/prowpresubmitsdataclusterstack-prowbucket7c73355c-vfwwxd2eb4gp/pr-logs/pull/aws_eks-anywhere/376/eks-anywhere-e2e-presubmit/1445473385474691072) | `/test eks-anywhere-e2e-presubmit`
eks-anywhere-presubmit | c9f464153c75b4bf74f1dd65de97f5c2da8b1e3d | [link](https://prow.eks.amazonaws.com/view/s3/prowpresubmitsdataclusterstack-prowbucket7c73355c-vfwwxd2eb4gp/pr-logs/pull/aws_eks-anywhere/376/eks-anywhere-presubmit/1445473385533411328) | `/test eks-anywhere-presubmit`
eks-anywhere-mocks-presubmits | c9f464153c75b4bf74f1dd65de97f5c2da8b1e3d | [link](https://prow.eks.amazonaws.com/view/s3/prowpresubmitsdataclusterstack-prowbucket7c73355c-vfwwxd2eb4gp/pr-logs/pull/aws_eks-anywhere/376/eks-anywhere-mocks-presubmits/1445473385504051200) | `/test eks-anywhere-mocks-presubmits`

[Full PR test history](https://prow.eks.amazonaws.com/pr-history?org=aws&repo=eks-anywhere&pr=376). [Your PR dashboard](https://prow.eks.amazonaws.com/pr?query=is:pr+state:open+author:g-gaston).

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>

**Comment by abhay-krishna** [comment]
@g-gaston: The following tests **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
eks-anywhere-attribution-files-presubmit | c9f464153c75b4bf74f1dd65de97f5c2da8b1e3d | [link](https://prow.eks.amazonaws.com/view/s3/prowpresubmitsdataclusterstack-prowbucket7c73355c-vfwwxd2eb4gp/pr-logs/pull/aws_eks-anywhere/376/eks-anywhere-attribution-files-presubmit/1445473385445330944) | `/test eks-anywhere-attribution-files-presubmit`
eks-anywhere-e2e-presubmit | c9f464153c75b4bf74f1dd65de97f5c2da8b1e3d | [link](https://prow.eks.amazonaws.com/view/s3/prowpresubmitsdataclusterstack-prowbucket7c73355c-vfwwxd2eb4gp/pr-logs/pull/aws_eks-anywhere/376/eks-anywhere-e2e-presubmit/1445473385474691072) | `/test eks-anywhere-e2e-presubmit`
eks-anywhere-presubmit | c9f464153c75b4bf74f1dd65de97f5c2da8b1e3d | [link](https://prow.eks.amazonaws.com/view/s3/prowpresubmitsdataclusterstack-prowbucket7c73355c-vfwwxd2eb4gp/pr-logs/pull/aws_eks-anywhere/376/eks-anywhere-presubmit/1445473385533411328) | `/test eks-anywhere-presubmit`
eks-anywhere-mocks-presubmits | a58190d86420ecfc34a2de15e8682e7ea295685f | [link](https://prow.eks.amazonaws.com/view/s3/prowpresubmitsdataclusterstack-prowbucket7c73355c-vfwwxd2eb4gp/pr-logs/pull/aws_eks-anywhere/376/eks-anywhere-mocks-presubmits/1445473999961198592) | `/test eks-anywhere-mocks-presubmits`

[Full PR test history](https://prow.eks.amazonaws.com/pr-history?org=aws&repo=eks-anywhere&pr=376). [Your PR dashboard](https://prow.eks.amazonaws.com/pr?query=is:pr+state:open+author:g-gaston).

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>

**Comment by abhay-krishna** [comment]
@g-gaston: The following tests **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
eks-anywhere-attribution-files-presubmit | c9f464153c75b4bf74f1dd65de97f5c2da8b1e3d | [link](https://prow.eks.amazonaws.com/view/s3/prowpresubmitsdataclusterstack-prowbucket7c73355c-vfwwxd2eb4gp/pr-logs/pull/aws_eks-anywhere/376/eks-anywhere-attribution-files-presubmit/1445473385445330944) | `/test eks-anywhere-attribution-files-presubmit`
eks-anywhere-presubmit | c9f464153c75b4bf74f1dd65de97f5c2da8b1e3d | [link](https://prow.eks.amazonaws.com/view/s3/prowpresubmitsdataclusterstack-prowbucket7c73355c-vfwwxd2eb4gp/pr-logs/pull/aws_eks-anywhere/376/eks-anywhere-presubmit/1445473385533411328) | `/test eks-anywhere-presubmit`
eks-anywhere-mocks-presubmits | a58190d86420ecfc34a2de15e8682e7ea295685f | [link](https://prow.eks.amazonaws.com/view/s3/prowpresubmitsdataclusterstack-prowbucket7c73355c-vfwwxd2eb4gp/pr-logs/pull/aws_eks-anywhere/376/eks-anywhere-mocks-presubmits/1445473999961198592) | `/test eks-anywhere-mocks-presubmits`
eks-anywhere-e2e-presubmit | a58190d86420ecfc34a2de15e8682e7ea295685f | [link](https://prow.eks.amazonaws.com/view/s3/prowpresubmitsdataclusterstack-prowbucket7c73355c-vfwwxd2eb4gp/pr-logs/pull/aws_eks-anywhere/376/eks-anywhere-e2e-presubmit/1445473999923449856) | `/test eks-anywhere-e2e-presubmit`

[Full PR test history](https://prow.eks.amazonaws.com/pr-history?org=aws&repo=eks-anywhere&pr=376). [Your PR dashboard](https://prow.eks.amazonaws.com/pr?query=is:pr+state:open+author:g-gaston).

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>

**Comment by abhay-krishna** [comment]
@g-gaston: The following tests **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
eks-anywhere-presubmit | c9f464153c75b4bf74f1dd65de97f5c2da8b1e3d | [link](https://prow.eks.amazonaws.com/view/s3/prowpresubmitsdataclusterstack-prowbucket7c73355c-vfwwxd2eb4gp/pr-logs/pull/aws_eks-anywhere/376/eks-anywhere-presubmit/1445473385533411328) | `/test eks-anywhere-presubmit`
eks-anywhere-mocks-presubmits | a58190d86420ecfc34a2de15e8682e7ea295685f | [link](https://prow.eks.amazonaws.com/view/s3/prowpresubmitsdataclusterstack-prowbucket7c73355c-vfwwxd2eb4gp/pr-logs/pull/aws_eks-anywhere/376/eks-anywhere-mocks-presubmits/1445473999961198592) | `/test eks-anywhere-mocks-presubmits`
eks-anywhere-e2e-presubmit | a58190d86420ecfc34a2de15e8682e7ea295685f | [link](https://prow.eks.amazonaws.com/view/s3/prowpresubmitsdataclusterstack-prowbucket7c73355c-vfwwxd2eb4gp/pr-logs/pull/aws_eks-anywhere/376/eks-anywhere-e2e-presubmit/1445473999923449856) | `/test eks-anywhere-e2e-presubmit`
eks-anywhere-attribution-files-presubmit | a58190d86420ecfc34a2de15e8682e7ea295685f | [link](https://prow.eks.amazonaws.com/view/s3/prowpresubmitsdataclusterstack-prowbucket7c73355c-vfwwxd2eb4gp/pr-logs/pull/aws_eks-anywhere/376/eks-anywhere-attribution-files-presubmit/1445473999889895424) | `/test eks-anywhere-attribution-files-presubmit`

[Full PR test history](https://prow.eks.amazonaws.com/pr-history?org=aws&repo=eks-anywhere&pr=376). [Your PR dashboard](https://prow.eks.amazonaws.com/pr?query=is:pr+state:open+author:g-gaston).

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>

**Review by vivek-koppuru — APPROVED** [review]
/lgtm

**Comment by abhay-krishna** [comment]
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/aws/eks-anywhere/pull/376#pullrequestreview-771626846" title="Approved">danbudris</a>*, *<a href="https://github.com/aws/eks-anywhere/pull/376#" title="Author self-approved">g-gaston</a>*, *<a href="https://github.com/aws/eks-anywhere/pull/376#pullrequestreview-771948419" title="LGTM">vivek-koppuru</a>*

The full list of commands accepted by this bot can be found [here](https://go.k8s.io/bot-commands?repo=aws%2Feks-anywhere).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[OWNERS](https://github.com/aws/eks-anywhere/blob/main/OWNERS)~~ [danbudris,g-gaston,vivek-koppuru]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

**Pull request merged**
